### PR TITLE
[crater experiment] Enable validate-mir by default

### PR DESCRIPTION
**Pull request by saethlin**
https://github.com/rust-lang/rust/pull/106612 made our MIR validation checks pickier. So this branch enables validation after every pass, to exercise those checks (and MIR validation in general) as much as possible.

Additionally, we recently merged https://github.com/rust-lang/rust/pull/106294 which comes with this remark:

> This [the UB this PR weaponizes] is hopefully rare enough to not break anything.

It would be good to know if these changes are causing issues in the ecosystem sooner rather than later.

Just a crater build run would exercise the MIR checks, but I would also like to see if crater can tell us anything about the `noundef` change. I think if we can do a crater test run with `+rustflags=-Copt-level=3` we have a chance of breaking tests if someone is accidentally relying on UB.

---

_sigh_ I forgot to r? @ghost, sorry about that. If someone wants to cruise through and fixup the labels/etc please do.

### CODE REVIEW

**Comment by jyn514** [comment]
@craterbot run start=master#159ba8a92c9e2fa4121f106176309521f4af87e9+rustflags=-Copt-level=3 end=try#613d1972a772987ed4ee06faccbab0e2fc839871+rustflags=-Copt-level=3 mode=build-and-test

**Comment by saethlin** [comment]
Well, I'm very glad we did this. Looks like we have a lot of legitimate regressions. I'll start triaging them in an hour or two and update this PR with my assessment.

**Comment by matthiaskrgr** [comment]
for a quick peak: `rg -o "broken MIR.*$"  |  awk 'NF=5'  | sort -n  | uniq`

**Comment by saethlin** [comment]
I haven't reviewed _all_ of the test-failed crates, but of those I have reviewed they are all flaky tests not UB. So I'm comfortable saying the `noundef` change isn't causing obvious widespread issues. That's nice.

At least two of the ICEs bisect to this:

searched nightlies: from nightly-2022-11-01 to nightly-2023-01-21
regressed nightly: nightly-2022-11-20
searched commit range: https://github.com/rust-lang/rust/compare/b833ad56f46a0bbe0e8729512812a161e7dae28a...c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
regressed commit: https://github.com/rust-lang/rust/commit/2f8d8040166a730d0da7bba0f2864f0ef7ff6364

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.5</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start 2022-11-01 --end 2023-01-21 --script script 
```
</details>

If I had to guess, this PR is causing these validation failures: https://github.com/rust-lang/rust/pull/104411 but that might just be because it is improving the validator.

I do however notice in the `pi_ecs` example, we get this error (formatted by hand):
```
Field projection `(*_1).field[0]` specified type
`pi_ecs::storage::SecondaryMap<pi_ecs::storage::LocalVersion, Node2Index>`,
but actual type is
`<Node2Index as pi_ecs::component::Component>::Storage`
```
And in `pi_ecs`: https://github.com/GaiaWorld/pi_ecs/blob/e46bb571032376d13b2cedf770291654f7e2d13f/src/component.rs#L23-L29
```
pub trait Component: ThreadSync + 'static {
	type Storage: Map<Key = LocalVersion, Val = Self> + Index<LocalVersion, Output=Self> + IndexMut<LocalVersion, Output=Self> + ThreadSync + 'static;
}

impl<C> Component for C where C: ThreadSync + 'static {
	default type Storage = SecondaryMap<LocalVersion, Self>;
}
```
So basically I'm well over my head figuring out what's actually happening here, and we found something but it wasn't what we were looking for.

**Comment by saethlin** [comment]
The ICE in `safer_ffi` bisects to:

searched nightlies: from nightly-2023-01-01 to nightly-2023-01-22
regressed nightly: nightly-2023-01-20
searched commit range: https://github.com/rust-lang/rust/compare/333ee6c466972185973d5097f8b5fb0f9fb13fa5...4c83bd03a9d94af35c97a6b8b595d40e291af84a
regressed commit: https://github.com/rust-lang/rust/commit/8b11574ca07f95dd33861d145a43708c20cb4eb5

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.5</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start 2023-01-01 --end 2023-01-22 --script script --regress=ice 
```
</details>

cc @Nilstrieb

**Comment by saethlin** [comment]
I do not know enough about the above issues to write up competent issues for them. I'll leave this open for a few more days and try to recruit some help.

**Comment by Nilstrieb** [comment]
The ICE in `safer_ffi` is expected and #106191, which is currently blocked on me and the lang team

**Comment by Nilstrieb** [comment]
Actually, looking at the safer-ffi logs it's also full of broken MIR? This should _really_ not be caused by my change

**Comment by saethlin** [comment]
Going to close this for cleanliness but try to keep it in my work queue to come back to if the above issue is fixed.
